### ISSUE TITLE
Function as an object property can trigger EVAL_STRING_ARG warning

### ISSUE THREAD

**Issue by scholtzm** [mozilla/addons-linter#1126]
### Describe the problem and steps to reproduce it:

1. Run `addons-linter` on code which contains nested function reference in `setTimeout` / `setInterval`.

### What happened?

Function which is nested inside object and referenced in `setTimeout` or `setInterval` will trigger `EVAL_STRING_VAL` warning.

### What did you expect to happen?

Warning should not be emitted at all.

### Anything else we should know?

Here's an example:

```js
var foo = {
  bar: function() {}
};

function foobar() {}

// EVAL_STRING_ARG
window.setTimeout(foo.bar);

// OK
window.setTimeout(foobar);

// OK
window.setTimeout(function() {
  console.log('foobar');
});
```

You can paste this into a file and then run `addons-linter .`. You will receive some errors regarding missing `manifest.json`, but most importantly this warning:

![image](https://cloud.githubusercontent.com/assets/2640934/22974577/4ee7b26c-f383-11e6-8449-abf7e9edb5a6.png)

**Comment by tofumatt** [comment]
Good catch, thanks for filing. This is probably to do with how we check for the type of things passed to `setInterval`, etc.

I _think_ the way we do (going off memory) is that we allow things declared as functions to be passed in and it's possible properties that are functions aren't seen as such by our JS checks.

As warnings will not cause validation to be rejected (but will cause reviewers to note the code) fortunately this will still be accepted on AMO.

We should at least change this error message though; I think any time we pass variables to setTimeout or setInterval it's worth at least noting, but this is at least the wrong message as it's not a string 😄

**Comment by scholtzm** [comment]
> we allow things declared as functions to be passed in and it's possible properties that are functions aren't seen as such by our JS checks.

Indeed. Here are the relevant bits of code:

https://github.com/mozilla/addons-linter/blob/master/src/rules/javascript/eval_string_arg.js#L13-L23
https://github.com/mozilla/addons-linter/blob/master/src/utils.js#L116-L130

**Comment by EnTeQuAk** [comment]
This will be fixed once https://github.com/mozilla/addons-linter/issues/1197 lands and replaces that rule.
